src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -383,7 +383,7 @@ private static string GenerateUpgradeReport(string? projectName, string backupPa
                     TelemetryServiceUtility.StartOrResumeTimer();
 
                     IReadOnlyList<ProjectAction> actions = await resolveActionsAsync(projectManagerService);
-                    IReadOnlyList<PreviewResult> results = await GetPreviewResultsAsync(projectManagerService, actions, cancellationToken);
+                    IReadOnlyList<PreviewResult> results = await GetPreviewResultsAsync(projectManagerService, actions, userAction, uiService, cancellationToken);
```

Comment:
What does not mockable mean?

It's a class with content. YOu can preload it with different data and pass it down. Isn't that sufficient?

Reply:
> What does not mockable mean?

Mockable is defined here: https://github.com/moq/moq/wiki/FAQ#moq---non-overridable-members-may-not-be-used-in-setup--verification-expressions

> It's a class with content. YOu can preload it with different data and pass it down. Isn't that sufficient?

No, because as I mentioned, I'm testing # calls to the object, not the values of the object.

Reply:
I'm not I follow why the call count is relevant and why it needs to be mockable. 

There's also a point in the doc you linked: 

> In this case there's no need to mock **PagingOptions **because it's easy to use a real one.

I think PackageSourceMapping is one of those classes where it's easy to use a real one.

Reply:
I'd be happy to learn how to count method calls without using Mocks. However, since I've created a follow-up issue, can we discuss that in that issue? If ultimately it needs to be closed, that's fine. It's really not relevant to this PR, though.